File: tricking_bees_python/models/player.py
from dataclasses import dataclass, field

from .card import Card, CardColor
from .card_stack import CardStack
from .role import Role


@dataclass
class Player:
    name: str
    tricks_won: int = 0
    hand: CardStack = field(init=False)
    current_role: Role = field(default_factory=Role)
    total_points: int = 0

    def __post_init__(self):
        self.hand = CardStack()

    def reset_for_new_subgame(self):
        """Resets the player for a new subgame."""
        self.tricks_won = 0
        self.hand = CardStack()
        self.current_role = Role()

    def deal_hand(self, card_stack: CardStack):
        """Deal a hand to the player."""
        self.hand = card_stack

    def try_play_card(
        self,
        card: Card,
        compulsory_color: CardColor | None = None,
    ) -> bool:
        """Try to play a card from the player's hand."""
        if card not in self.hand.cards:
            return False
        if compulsory_color is not None and self.hand.contains_color(compulsory_color):
            if card.color != compulsory_color:
                return False
        self.hand.cards.remove(card)
        return True

    def get_playable_cards(
        self, has_turn: bool = False, compulsory_color: CardColor | None = None
    ) -> list[Card]:
        """Return the cards the player can currently play."""
        if not has_turn:
            return []
        return self.current_role.get_playable_cards(self.hand, compulsory_color)

File: tricking_bees_flutter/packages/tb_core/lib/src/models/player.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:yust/yust.dart';

import '../cards/card_color.dart';
import '../cards/card_stack.dart';
import '../cards/game_card.dart';
import '../roles/role.dart';
import '../roles/role_catalog.dart';
import 'game/game.dart';
import 'game/logging/points_awarded.dart';

part 'player.g.dart';

@JsonSerializable()

/// A player of the game.
class Player {
  /// Creates a [Player].
  Player({
    required this.id,
    required this.displayName,
    CardStack? cards,
    this.tricksWon = 0,
    this.pointTotal = 0,
    RoleCatalog? roleKey,
  })  : roleKey = roleKey ?? RoleCatalog.noRole,
        cards = cards ?? CardStack();

  /// Creates a dummy [Player].
  factory Player.dummy({CardStack? cards}) => Player(
        id: 'dummy',
        displayName: 'dummy',
        cards: cards,
      );

  /// Creates a [Player] from a [YustUser].
  factory Player.fromUser(YustUser user) => Player(
        id: user.id,
        displayName: user.firstName,
      );

  /// Creates a [Player] from a JSON map.
  factory Player.fromJson(Map<String, dynamic> json) => _$PlayerFromJson(json);

  /// Converts the [Player] to a JSON map.
  Map<String, dynamic> toJson() => _$PlayerToJson(this);

  /// The id of the player.
  final String id;

  /// The name of the player.
  final String displayName;

  /// The amount of tricks the player has won.
  int tricksWon;

  /// The cards the player currently holds in their hand.
  CardStack cards;

  /// The role the player has currently chosen.
  RoleCatalog roleKey;

  /// The amount of points this player has achieved.
  int pointTotal;

  /// An empty [Player].
  static final empty = Player(
    id: 'EMPTY',
    displayName: 'EMPTY',
  );

  /// The role of the player.
  Role get role => Role.fromRoleCatalog(roleKey);

  /// Awards points to this player.
  void awardPoints(Game game, int playerIndex) {
    final points = calculateCurrentPoints(game);
    pointTotal += points;
    game.addLogEntry(
      LogPointsAwarded(
        playerIndex: playerIndex,
        roleKey: roleKey,
        points: points,
        tricksWon: tricksWon,
      ),
      round: game.currentRound - 1,
    );
  }

  /// Get this player ready for the new subgame.
  void resetForNewSubgame() {
    tricksWon = 0;
    cards = CardStack();
    roleKey = RoleCatalog.noRole;
  }

  /// Deal this player the given cards.
  void dealCards(CardStack newCards) {
    cards.addCards(newCards);
  }

  /// Returns whether the player can play a given card, given it's their turn.
  bool canPlayCard(GameCard card, CardColor? compulsoryColor) =>
      role.getPlayableCards(cards, compulsoryColor).contains(card);

  /// Play the given card.
  void playCard(GameCard card) {
    cards.removeCard(card);
  }

  /// Calculates the points this player achieved in the current subgame.
  int calculateCurrentPoints(Game game) =>
      role.calculatePoints(game, tricksWon);
}
